ec2_vpc_route_table_info upgraded to boto3

prasadkatti opened the pull request:
##### SUMMARY
ec2_vpc_route_table_info upgraded to boto3

This is #44966 rebased onto devel
Fixes #63112

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ec2_vpc_route_table_info

##### ADDITIONAL INFORMATION
This PR is just a rebase of #44966 onto the latest devel
This is my first time rebasing someone else's PR. So let me know if I need to change something.

Commit:
ec2_vpc_route_table_facts.py migrated to boto3

mpostument committed:
added route_table_ids in description

mpostument committed:
version changed to 2.7

mpostument committed:
fix version added error

mpostument committed:
fix comments to PR

prasadkatti committed:
ec2_vpc_route_table_info version_added changed to 2.9

prasadkatti commented:
What should be the version_added field for this module? The test is failing because it is looking for '2.0'. Also what should be version_added for the new option to the module?

tremble commented:
version added should be 2.10

Unfortunately 2.9 has now been branched and is in a feature freeze pending it's release "soon".

tremble commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -52,81 +57,174 @@
 
 
 def main():
     argument_spec = ec2_argument_spec()
```
AnsibleAWSModule automatically adds ec2_argument_spec() so you can simplify 213 through 219

```
    argument_spec = dict(
        route_table_ids=dict(default=[], type='list'),
        filters=dict(default={}, type='dict')
    )

tremble commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -52,81 +57,174 @@
-
-
-def get_route_table_info(route_table):
+    pass  # caught by imported HAS_BOTO3
```
Actually caught by AnsibleAWSModule

prasadkatti committed:
[ec2_vpc_route_table_info] Fixed version_added values

tremble commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -18,13 +18,18 @@
+    description:
+      - Get details of a specific route tables. This value should be provided as a list.
+    required: false
+    version_added: "2.9"
```
```suggestion
    version_added: "2.10"
```

tremble reviewed (changes requested):
Especially given that this is a boto -> boto3 migration it would be good to add some integration tests.

we already have test/integration/targets/ec2_vpc_route_table/ adding `ec2_vpc_route_table_info` to the aliases file and adding ec2_vpc_route_table_info tests in there too would be really good.

prasadkatti committed:
add attribute id for route table

prasadkatti committed:
ec2_vpc_route_table_info fixes for review comments

prasadkatti commented:
This is ready for review. But I have added some integration tests for this module in #63413. It would be good to kick off the CI on this PR after the other one has been merged. The PR with the tests is waiting for ansibot to merge it in.

prasadkatti commented:
ready_for_review

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -25,6 +25,13 @@
+      - Get details of a specific route tables. This value should be provided as a list.
+    required: false
+    type: list
+    version_added: "2.10"
```
You could add `default: []` here (since omitting means it is `None`).

prasadkatti replied:
done.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -53,80 +60,174 @@
+                           ignore_list=['Tags']) for route_table in result['RouteTables']]
+    if snaked_route_tables:
+        for route_table in snaked_route_tables:
+            route_table['id'] = route_table['route_table_id']
```
This looks like one you added for backward compatibility - do you mind also adding it to the return documentation? It would also be good to have a test for the return values from the boto version as a precaution - you could add something like this to the ec2_vpc_route_table integration test after https://github.com/ansible/ansible/blob/devel/test/integration/targets/ec2_vpc_route_table/tasks/main.yml#L493-L511:
```
- name: test route table facts
  ec2_vpc_route_table_info:
    filters:
      route-table-id: "{{ route_table_id }}"
    <<: *aws_connection_info
  register: result

- assert:
    that:
      - result.id
      - etc
```

prasadkatti replied:
Sure. I will add it to the return documentation. I already added a test for checking if `id` is returned [here](https://github.com/ansible/ansible/pull/63413/files#diff-ea65d1c9ceeb31fc1b20355ec26f4dbfR593). I will add a test to check if `route_table_id` is also returned.
Is there any sort of deprecation policy to remove `id` from the returned values to stay in sync with boto3? Or are we now going to have it returned forever?

prasadkatti replied:
The PR for `ec2_vpc_route_table_info` was merged earlier today - #63413

prasadkatti replied:
Should I add the test case in a separate PR or should I rebase this on devel (so it includes #63413)?

s-hertel replied:
@prasadkatti Ideally the tests would be merged before porting to boto3, so if you don't mind making a second PR to get in first that would be awesome. I don't think a top level `id` is the only thing that changed between boto and boto3. Ideally there's be an assertion for each return value. I know that's tedious. `routes` and `associations` both seem to have more keys in the boto output here - https://github.com/ansible/ansible/pull/44966#issue-212382602 (that comment may be incomplete though, as there are no propagating_vgws - I don't know whether boto returned if they exist, and if so, whether those contents differ between boto and boto3 too).

We don't have a good way of deprecating return values at the moment, unfortunately. I left a comment here with some thoughts on that https://github.com/ansible/ansible/pull/62985#issuecomment-544986596. Feel free to weigh in if you have ideas!

s-hertel replied:
Also, thanks for the pointer to the PR merged earlier. I guess the `id` case is covered. :smile: Most of the others may be too, but if you could check that all of the return values are accounted for within reason that would be appreciated.

prasadkatti replied:
I have a question about this. The `route_tables.associations.id` (boto) has changed to `route_tables.associations.route_table_association_id` (boto3). Are we supposed to maintain backward compatibility for all such attributes? That sounds like a lot of work. I can try to do that if required. But can't we just add a release note for this module to let users know that some of the return values for this module have changed?

jillr replied:
We'll need a standard 4 release deprecation cycle to change the output, we always want to be sure we give people a fair chance to update their playbooks that may depend on return values or any other functionality. I haven't tested, but I would expect you should be able to set result.route_tables.associations.id to result.route_tables.associations.route_table_association_id before returning the result.  Then we would need the module RETURN docs and a porting guide entry to state that the return value is being deprecated.  #52518 has a good example of the needed docs.

prasadkatti replied:
Thanks for the pointers. I am trying to gather a list of differences between boto and boto3 return values. This is what I have so far -

1. `route_tables.associations.id` is now `route_tables.associations.route_table_association_id`
2. Some fields seem to be returned conditionally in boto3. Before, in boto, they were always returned and would have a null value wherever applicable. Examples include `route_tables.associations.subnet_id`; `instance_id`, `interface_id` and `vpc_peering_connection_id` in `route_tables.routes`.
3. `propagating_vgws` - seems like a new return value

I know what needs to be done for (1). Don't think any action is needed for (3). For (2), do we want to add the missing keys with a null value?

Also, the above list is the difference that I could find. There might be more.

jillr replied:
I think it's better to set them to null if they always existed before - unknown key would behave differently than a key that exists but is set to a null value and we don't want to break things for people if possible. Thanks very much for diving in to this.

tremble replied:
@prasadkatti  At this point I think the remaining work is
a) set the values to None where boto would have done.  I don't particularly like this, but doing otherwise will break things for users.  The worst thing is we can't test for people trying to use it and spit out warnings.

b) Add some test cases for the return values.  test/integration/targets/ec2_vpc_route_table/tasks/main.yml already includes all of the framework so this should be relatively simple...

prasadkatti committed:
ec2_vpc_route_table_info - doc fixes from review comments

- add a default value for route_table_ids in documentation
- add id to the return documentation

prasadkatti committed:
maintain backwards compatibility with boto

tremble commented:
Hi @prasadkatti I think we were pretty close to getting this merged.  Unfortunately having cleared out the sanity errors I introduced a merge conflict here.  Should be quick and easy to fix.

prasadkatti commented:
@tremble - I have resolved the merge conflicts.

tremble commented:
@prasadkatti The changes need rebasing rather than merging.

prasadkatti committed:
add tests cases to ensure backwards compatibility with boto

prasadkatti commented:
@tremble - this is ready for review. I picked this task up, but I am not super familiar with VPCs and networking. The last commit made to maintain backwards compatibility with boto definitely needs to be reviewed carefully.

tremble reviewed (commented):
Generally looks good.

I'd like to see more tests, specifically around the compatibility pieces.  If you're doing what I think you're doing with the JSON round-tripping then it would be good to validate some of the actual return values rather than just "does this key exist".

jillr commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -54,80 +61,187 @@
+              type: str
+            instance_owner_id:
+              description: The ID of a NAT instance in your VPC.
+              returned: when NAT instance is available
```
```suggestion
              description: The ID of a NAT instance owner in your VPC.
              returned: when NAT instance owner is available
```

jillr reviewed (commented):
One small docs note I missed previously, it would be nice to get this in before freeze on the 2nd if you are able to add the tests tremble mentioned - you've put a lot of work into it!

prasadkatti committed:
update description for instance_owner_id

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

tremble reviewed (approved):
From IRC:

> tremble>  jillr, https://github.com/ansible/ansible/pull/63168 I think the value of getting it in before the freeze trumps making sure the tests are perfect.

tremble commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -26,6 +26,13 @@
+    required: false
+    type: list
+    default: []
+    version_added: "2.10"
```
```suggestion
    version_added: "2.10"
    elements: str
```

jillr commented:
The sanity test failure will need to be addressed, once that is we're good to merge.

tremble commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -54,80 +61,187 @@
-            filters=dict(default=None, type='dict')
-        )
+    argument_spec = dict(
+        route_table_ids=dict(default=[], type='list'),
```
```suggestion
        route_table_ids=dict(default=[], type='list', elements='str'),
```

prasadkatti committed:
fix sanity failure

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

prasadkatti committed:
fix sanity failure

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

prasadkatti commented:
If this can wait till tomorrow, I can try to add the tests that @tremble asked for. Sorry about the delay.

prasadkatti commented on lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py:
```diff
@@ -54,80 +62,187 @@
+    if snaked_route_tables:
+        for route_table in snaked_route_tables:
+            route_table['id'] = route_table['route_table_id']
+            for association in route_table['associations']:
+                association['id'] = association['route_table_association_id']
+                if 'subnet_id' not in association:
+                    association['subnet_id'] = None
+            for route in route_table['routes']:
+                for key in ['instance_id']:
+                    if key not in route:
+                        route[key] = None
```
Hi @tremble, this is the part where we add some keys to the returned dict to make it backwards compatible. Boto always had these keys defined and they are set to None when necessary. Now in boto3, these keys are missing from the dict if they are not applicable. The 2 tests that I added ensure that we maintain backwards compatibility by always having those keys present in the returned dict. When these keys are added their values are set to None to mimic boto2 behavior. So what tests exactly do I need to add here?
